File: example_player.py
import sys
import socket
import random


def eprint(*args, **kwargs):
    print(*args, file=sys.stderr, **kwargs)


class GameState:
    def __init__(self, match_string):
        if match_string[:11] != "MATCHSTATE:":
            raise AssertionError("ERROR: message must starts with 'MATCHSTATE:'")
        match_string = match_string[11:]
        message = match_string.strip().split(":")

        # [<position>, <handNumber>, <betting>, <cards>]
        self.message = message
        # int, <position>
        self.viewingPlayer = int(message[0])
        # int, <handNumber>
        self.handId = int(message[1])
        # string, <betting>
        self.betting_string = message[2]
        # list, each element is a string of cards dealt in each round.
        # e.g. ['9hQd']: 9hQd are dealt as private cards in the first round
        #      ['5d5c', '8dAs8s', '4h']: 9hQd are dealt as private cards in the first round,
        #           8dAs8s are dealt in the second round, and 4h is dealt in the third round.
        # NOTICE: do not use this when the game is finished.
        self.cards = self._cal_card_sequence(message[3])

        # A nested list of shape
        # (number of rounds played, number of actions in that round).
        # Actions are "c", "f", "r300"(or any other valid betting size)
        self.betting_sequence = self._cal_betting_sequence(self.betting_string)
        # Current round. One of (0, 1, 2, 3).
        self.round = len(self.betting_sequence)

        # bool, whether the hand is finished
        self.finished = None
        # bool, whether the agent should act
        self.acting = None
        # list of two ints. e.g. [100, 50] means player at position 0 has spent
        # 100, and player at position 1 has spent 50.
        self.spent = None

        # Calculate self.finished, self.acting, and self.spent.
        self._simulate_game()

    @staticmethod
    def _cal_betting_sequence(betting_string):
        betting_sequence = list()
        for x in betting_string.split("/"):
            sequence = list()
            s = 0
            while s < len(x):
                if x[s] == 'c' or x[s] == 'f':
                    sequence.append(x[s])
                    s += 1
                else:
                    e = s + 1
                    while e < len(x) and \
                            (x[e] != 'c' and x[e] != 'f' and x[e] != 'r'):
                        e += 1
                    sequence.append(x[s:e])
                    s = e
            betting_sequence.append(sequence)
        return betting_sequence
    
    @staticmethod
    def _cal_card_sequence(cards_string):
        cards = cards_string.split("/")
        if cards[0][0] == "|":
            cards[0] = cards[0][1:]
        else:
            cards[0] = cards[0][:-1]
        return cards

    def _simulate_game(self):
        finished = False
        spent = [100, 50]   # big blind & small blind
        acting = 1
        for r, s in enumerate(self.betting_sequence):
            if r > 0:
                acting = 0
            acted = 0
            for a in s:
                if a[0] == 'f':
                    finished = True
                elif a[0] == 'c':
                    spent[acting] = spent[1-acting]
                    acted += 1
                    if acted >= 2 and r == 3:
                        finished = True
                else:
                    spent[acting] = int(a[1:])
                    acted += 1
                acting = 1 - acting

        if spent[self.viewingPlayer] == 20000:
            finished = True
        self.finished = finished
        self.acting = (not finished) and (acting == self.viewingPlayer)
        self.spent = spent

    def is_fold_valid(self) -> bool:
        """Check whether fold is a valid action in current state.

        :return: True if fold is valid, False otherwise.
        """
        if self.spent[0] == self.spent[1]:
            return False
        return True

    def is_raise_valid(self):
        """Check if raise if a valid action in current state. If valid, give
        min and max raise bets (the amount that can be raised TO, NOT raised
        BY).

        :return: None if raise is not valid, (min_bet, max_bet) otherwise.
        """
        if self.spent[1-self.viewingPlayer] == 20000:
            return None
        spent = [100, 50]
        acting = 1
        for r, s in enumerate(self.betting_sequence[:-1]):
            if r > 0:
                acting = 0
            for a in s:
                if a[0] == 'c':
                    spent[acting] = spent[1-acting]
                else:
                    spent[acting] = int(a[1:])
                acting = 1 - acting
        if len(self.betting_sequence) > 1:
            assert(spent[0] == spent[1])
        else:
            assert(spent == [100, 50])

        current_bet = spent[0]
        min_raise = 100
        for a in self.betting_sequence[-1]:
            if a[0] == 'r':
                min_raise = int(a[1:]) - current_bet
                current_bet = int(a[1:])
        return min(current_bet + min_raise, 20000), 20000


def act(state: GameState):
    a = random.randint(0, 3)
    if a == 1 and state.is_fold_valid():
        return "f"
    elif (a == 2 or a == 3) and state.is_raise_valid():
        min_bet, max_bet = state.is_raise_valid()
        if a == 2:
            return "r" + str(min_bet)
        else:
            return "r" + str(max_bet)
    else:
        return "c"


if __name__ == "__main__":
    if len(sys.argv) < 3:
        eprint("usage: player server port")
        sys.exit(1)

    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.connect((sys.argv[1], int(sys.argv[2])))

    if sock.send("VERSION:2.0.0\n".encode()) != 14:
        eprint("ERROR: could not send version to server")
        sys.exit(1)
    
    while True:
        lines = sock.recv(4096).decode()  # we may read multiple lines at a time
        if lines is None or len(lines) <= 0:
            break
        for line in lines.strip().split("\r\n"):
            game_state = GameState(line)
            if game_state.acting is True:
                action = act(game_state)
                ret = line + ":"
                ret += action
                ret += "\r\n"
                if sock.send(ret.encode()) != len(ret):
                    eprint("ERROR: could not send response to server")
                    sys.exit(1)

    sock.close()
    sys.exit(0)

File: example_player.cpp
#include <stdlib.h>
#include <stdio.h>
#include <assert.h>
#include <string.h>
#include <unistd.h>
#include <netdb.h>
#include <sys/socket.h>
#include <sys/time.h>
#include <netinet/in.h>
#include <netinet/tcp.h>
#include <getopt.h>
#include "game.h"
#include "rng.h"
#include "net.h"


/* TODO: implement your own poker strategy in this function!
 * 
 * The function is called when it is the agent's turn to play!
 * All the information about the game and the current state is preprocessed
 * and stored in game and state.
 * 
 * */
Action act(Game *game, MatchState *state, rng_state_t *rng) {
  Action action;

  // TODO: substitute this simple random strategy with your own strategy

  double probs[ NUM_ACTION_TYPES ];
  double actionProbs[ NUM_ACTION_TYPES ];
  /* Define the probabilities of actions for the player */
  probs[ a_fold ] = 0.06;
  probs[ a_call ] = ( 1.0 - probs[ a_fold ] ) * 0.5;
  probs[ a_raise ] = ( 1.0 - probs[ a_fold ] ) * 0.5;

  /* build the set of valid actions */
  double p = 0.0;
  int a;
  for( a = 0; a < NUM_ACTION_TYPES; ++a ) {
    actionProbs[ a ] = 0.0;
  }

  /* consider fold */
  action.type = a_fold;
  action.size = 0;
  if( isValidAction( game, &(state->state), 0, &action ) ) {
    actionProbs[ a_fold ] = probs[ a_fold ];
    p += probs[ a_fold ];
  }

  /* consider call */
  action.type = a_call;
  action.size = 0;
  actionProbs[ a_call ] = probs[ a_call ];
  p += probs[ a_call ];

  /* consider raise */
  int32_t min, max;
  if( raiseIsValid( game, &(state->state), &min, &max ) ) {
    actionProbs[ a_raise ] = probs[ a_raise ];
    p += probs[ a_raise ];
  }

  /* normalise the probabilities  */
  assert( p > 0.0 );
  for( a = 0; a < NUM_ACTION_TYPES; ++a ) {

    actionProbs[ a ] /= p;
  }

  /* choose one of the valid actions at random */
  p = genrand_real2( rng );
  for( a = 0; a < NUM_ACTION_TYPES - 1; ++a ) {
    if( p <= actionProbs[ a ] ) {
      break;
    }
    p -= actionProbs[ a ];
  }
  action.type = (enum ActionType)a;
  if( a == a_raise ) {
    action.size = min + genrand_int32( rng ) % ( max - min + 1 );
  }
  else {
    action.size = 0;
  }

  return action;
}


/* Anything related with socket is handled below. */
/* If you are not interested with protocol details, you can safely skip these. */

int step(int len, char line[], Game *game, MatchState *state, rng_state_t *rng) {
  /* add a colon (guaranteed to fit because we read a new-line in fgets) */
  line[ len ] = ':';
  ++len;

  Action action = act(game, state, rng);

  /* do the action! */
  assert( isValidAction( game, &(state->state), 0, &action ) );
  int r = printAction( game, &action, MAX_LINE_LEN - len - 2, &line[ len ] );
  if( r < 0 ) {
    fprintf( stderr, "ERROR: line too long after printing action\n" );
    exit( EXIT_FAILURE );
  }
  len += r;
  line[ len ] = '\r';
  ++len;
  line[ len ] = '\n';
  ++len;

  return len;
}


int main( int argc, char **argv )
{
  int sock, len;
  uint16_t port;
  Game *game;
  MatchState state;
  FILE *file, *toServer, *fromServer;
  struct timeval tv;
  rng_state_t rng;
  char line[ MAX_LINE_LEN ];

  /* we make some assumptions about the actions - check them here */
  assert( NUM_ACTION_TYPES == 3 );

  if( argc < 3 ) {

    fprintf( stderr, "usage: player server port\n" );
    exit( EXIT_FAILURE );
  }

  /* Initialize the player's random number state using time */
  gettimeofday( &tv, NULL );
  init_genrand( &rng, tv.tv_usec );

  /* get the game */
  file = fopen( "./holdem.nolimit.2p.reverse_blinds.game", "r" );
  if( file == NULL ) {

    fprintf( stderr, "ERROR: could not open game './holdem.nolimit.2p.reverse_blind.game'\n");
    exit( EXIT_FAILURE );
  }
  game = readGame( file );
  if( game == NULL ) {

    fprintf( stderr, "ERROR: could not read game './holdem.nolimit.2p.reverse_blind.game'\n");
    exit( EXIT_FAILURE );
  }
  fclose( file );

  /* connect to the dealer */
  if( sscanf( argv[ 2 ], "%"SCNu16, &port ) < 1 ) {

    fprintf( stderr, "ERROR: invalid port %s\n", argv[ 2 ] );
    exit( EXIT_FAILURE );
  }
  sock = connectTo( argv[ 1 ], port );
  if( sock < 0 ) {

    exit( EXIT_FAILURE );
  }
  toServer = fdopen( sock, "w" );
  fromServer = fdopen( sock, "r" );
  if( toServer == NULL || fromServer == NULL ) {

    fprintf( stderr, "ERROR: could not get socket streams\n" );
    exit( EXIT_FAILURE );
  }

  /* send version string to dealer */
  if( fprintf( toServer, "VERSION:%"PRIu32".%"PRIu32".%"PRIu32"\n",
	       VERSION_MAJOR, VERSION_MINOR, VERSION_REVISION ) != 14 ) {

    fprintf( stderr, "ERROR: could not get send version to server\n" );
    exit( EXIT_FAILURE );
  }
  fflush( toServer );

  /* play the game! */
  while( fgets( line, MAX_LINE_LEN, fromServer ) ) {

    /* ignore comments */
    if( line[ 0 ] == '#' || line[ 0 ] == ';' ) {
      continue;
    }

    len = readMatchState( line, game, &state );
    if( len < 0 ) {

      fprintf( stderr, "ERROR: could not read state %s", line );
      exit( EXIT_FAILURE );
    }

    if( stateFinished( &state.state ) ) {
      /* ignore the game over message */

      continue;
    }

    if( currentPlayer( game, &state.state ) != state.viewingPlayer ) {
      /* we're not acting */

      continue;
    }

    len = step(len, line, game, &state, &rng);

    if( fwrite( line, 1, len, toServer ) != len ) {

      fprintf( stderr, "ERROR: could not get send response to server\n" );
      exit( EXIT_FAILURE );
    }
    fflush( toServer );
  }

  return EXIT_SUCCESS;
}
